### DISCUSSION TITLE
Generic atoms?

### DISCUSSION THREAD

**Discussion by rodrigocfd** [pmndrs/jotai#1760] [Q&A] [answered]
I'm experimenting with the boundaries of the atoms, and I just came up with this (which doesn't compile):

```ts
interface Foo<T> {
  name: string;
  thing: T;
}

const things = atom<T>([] as Foo<T>[]);
```

Or, maybe the simplest form:

```ts
const things = atom<T>([] as T[]);
```

Are generic atoms currently supported in some way?

**Comment by dai-shi** [comment]
atom is not a function, so it can't be generic, no?

**Reply by rodrigocfd** [reply]
Ouch, you're correct.

My goal was to have a `fetch` wrapper that could write to an atom, like:

```ts
const doGet = atom(null, async (get, set, url: string, obj: string): Promise<T> => {
  const resp = await fetch(url);
  const json = await resp.json();
  // write to atom...
  return json as T;
});
```

... but atoms are concrete values, so they obviously *cannot* be generic.

The solution to the above is to have specialized write-only atoms to each type, like `getPeopleAtom` and `getBikesAtom`, so they would know the type to work upon, without the need for the generic parameter.

Thank you.

**Comment by rodrigocfd** [comment, marked as answer]
I guess I wrote what I was looking for:

```ts
import {atom, useAtom} from 'jotai';

const nameAtom = atom('');

export function useGet(): <T>(url: string) => Promise<T> {
    const [name, setName] = useAtom(nameAtom);

    return async <T>(url: string): Promise<T> => {
        const resp = await fetch(url);
        const json = await resp.json();

        setName('STORE SOMETHING IN THE ATOM IF YOU WANT TO');

        return json as T;
    };
}
```

The hook above returns a function which calls `fetch`, returns a generic value and is capable of writing to atoms.

Usage:

```ts
import {useGet} from './atoms';

function App() {
    const doGet = useGet();

    async function clicked() {
        const data = await doGet<string>('/hello');
    }

    return <button onClick={clicked}>Click me</button>;
}

export default App;
```

My use case is that the `fetch` can receive errors/warnings from the server, and these must be globally displayed in a component at the top of the page. And this solution solves the problem perfectly.
